packages/store/src/index.ts

```diff
@@ -0,0 +1,231 @@
+  actions: undefined extends A ? undefined : A;
+  setState: SetStoreFunction<T>;
+};
+
+/**
+ * Factory to create reusable stores.
+ *
+ * @typedef StoreFactory
+ * @template <T: state type of the store, G: store access methods, A: store mutation methods>
+ * @returns (initialValueSetter?: T | ((fallbackState: T) => T)) => SharedStore<T, G, A>
+ */
+export type StoreFactory<T extends object, G = undefined, A = undefined> = (
+  initialValueSetter?: T | ((fallbackState: T) => T)
+) => SharedStore<T, G, A>;
+
+/**
+ * createSharedStore takes an `initialState` and an optional object containing functions to create getters and/or actions.
+ * A store is created and returned along with access and mutation functions & the native `setState`.
+ * @template <T: state type of the store, G: store access methods, A: store mutation methods>
+ * @param initialState object to be wrapped in a Store
+ * @param createMethods optional object containing functions to create getters and/or actions.
+ * @returns `SharedStore`
+ * @example
+ * ```tsx
+ * // `counter-store.ts`
+ * export const counterStore = createSharedStore({
+ *   value: 5,
+ * }, {
+ *  getters: state => ({
+ *   count: () => state.value,
+ *  }),
+ *  actions: (setState, state) => ({
+ *   increment: () => setState(val => ({ ...val, value: val.value + 1 })),
+ *   reset: () => setState("value", 0),
+ *  })
+ * });
+ *
+ * // `Example.tsx`
+ * import { counterStore } from './counter-store.ts';
+ *
+ * const {state: counterState, getters: {count}, actions: {increment, reset}} = counterStore;
+ * count() // => 5
+ * increment()
+ * count() // => 6
+ * reset() // => 0
+ * ```
+ */
+export function createSharedStore<
+  T extends object,
+  G extends object | undefined,
+  A extends object | undefined
+>(
+  initialState: T,
+  createMethods: {
+    getters: (state: T) => G;
+    actions: (setState: SetStoreFunction<T>, state: T) => A;
+  }
+): SharedStore<T, G, A>;
+export function createSharedStore<T extends object, G extends object | undefined>(
+  initialState: T,
+  createMethods: {
+    getters: (state: T) => G;
+  }
+): SharedStore<T, G>;
+export function createSharedStore<T extends object, A extends object | undefined>(
+  initialState: T,
+  createMethods: {
+    actions: (setState: SetStoreFunction<T>, state: T) => A;
+  }
+): SharedStore<T, undefined, A>;
+export function createSharedStore<T extends object>(
+  initialState: T,
+  createMethods?: {}
+): SharedStore<T>;
+export function createSharedStore<
+  T extends object,
+  G extends object | undefined,
+  A extends object | undefined
+>(
+  initialState: T,
+  createMethods?: {
+    getters?: (state: T) => G;
+    actions?: (setState: SetStoreFunction<T>, state: T) => A;
+  }
+) {
+  const [state, setState] = createStore<T>(initialState);
+  const getters = createMethods?.getters?.(state);
+  const actions = createActions(createMethods?.actions?.(setState, state));
+
+  return {
+    state,
+    getters,
+    actions,
+    setState
+  };
```

Comment:
```suggestion
  return {
    state,
    getters,
    actions
  };
```
I thinking if the setter is even supposed to be shared by default. It won't push people to read/write segregation if you have the setter always available. all the writes should be ideally covered by actions.
If you really want to have access to a setter, you can make it into an action:
```ts
actions: (setState, state) => ({
  set: setState,
}),
```

Reply:
Oh wow I was really over thinking that. Even after seeing your example I thought you meant to incorporate `set` as part of the store creation 😂